Add system for dimensions to be marked for deletion

diesieben07 opened the pull request:
When calling `DimensionManager.markForDeletion(DimensionType)` the `DimensionType` will not be re-created when the world is reloaded and the save-directory for the dimension will be cleaned up.

diesieben07 committed:
Add system for dimensions to be marked for deletion, which will not remove them and their save-data from the save upon reload

diesieben07 committed:
add debug mod

diesieben07 committed:
fix testmod

diesieben07 committed:
fix testmod and formatting

diesieben07 committed:
clear folders to remove on load

diesieben07 committed:
Remove unneeded method

diesieben07 committed:
undo build.gradle update

diesieben07 committed:
undo import changes

diesieben07 committed:
undo testmod change

diesieben07 committed:
small testmod improvement

diesieben07 committed:
Improve documentation

diesieben07 commented:
This PR comes with a test-mod that adds an Item.
Upon right-click this item creates a player-unique dynamic dimension (looks just like the overworld) and teleports the player there.
Upon leaving the dimension it is marked for deletion.

Shift-right-clicking the item prints the current dimension to chat.

diesieben07 committed:
improve testmod

McJty commented:
This PR is essential for mods like Mystcraft and RFTools Dimensions. Thanks for making it

50ap5ud5 commented:
Thank you for doing this, I think this would help fix a similar issue I was having. :)

LexManos commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -199,14 +209,49 @@ public static ServerWorld getWorld(MinecraftServer server, DimensionType dim, bo
+        }
+        getData(dim).markedForDeletion = true;
+
+        Path base = Paths.get(".").toAbsolutePath().normalize();
```
This is now how you get the current directory, we have FMLPaths which should have the supplied --gamedir paths.
Nothing in Minecrat should rely on the current execution directory directly. {Server's should set the gamedir to CWD if none is supplied}

diesieben07 replied:
It doesn't matter what the path is. This is simply to check that the dimension returns a proper path, and get that path relative to the save root.
I could use `Paths.get("foobar")` as well.

diesieben07 replied:
In fact `--gamedir` would be just as "wrong" here, it would have to be the save path, which is not available at this point, as far as I can tell.

LexManos commented on src/main/java/net/minecraftforge/common/DimensionManager.java:
```diff
@@ -408,6 +457,31 @@ public static void readRegistry(CompoundNBT data)
+                .forEach(foldersScheduledForDeletion::add);
+    }
+
+    public static void processScheduledDeletions(SaveHandler saveHandler)
```
Is there a way we can add a confirmation screen similiar to the registry issues one?
I don't like the idea of deleting files in bulk without atleast a confirmation.
Also, should filter this list with a folder.exists() so we don't warn we're deleting folders that are already gone.

RobertSkalko replied:
If that gets added then i'd want an option to opt out. The dimensions in my mod are supposed to be treated as scrappable timed items, not like registries. Would be very annoying to have that dialog every time you start the game again. 

I specifically warn players not to build in those dimensions and that they are timed.

Maybe the markForDeletition() could by default add the warning but have another same method with a boolean overload?

LexManos replied:
Then your mod setup is bad, Dimensions are large objects filled with user data. And should not be treated as lightweight trash objects.
You should rethink your design and see if its possible to get away with reusing the dimensions and just resetting the changes users do.

RobertSkalko replied:
I didn't mean treated as lightweight, i meant as something that shouldn't be saved.

Like instances of dungeons in mmorpgs. And yes i did use workarounds while the system wasn't there in 1.13 but it wasn't as good as fresh dimensions. But i don't want to discuss workarounds in this PR, i can do that in discord if you need.

LexManos replied:
Pretty sure we talked about this on Discord, Your usecase is a bad one. Throwing away entire dimensions is not a good system.

diesieben07 committed:
Merge branch '1.15.x' into mark-dimension-for-deletion

# Conflicts:
#	src/main/java/net/minecraftforge/common/DimensionManager.java

diesieben07 committed:
Add startupquery before deleting dimensions

diesieben07 commented:
Fixed for new changes on target branch.

I also added a confirmation via `StartupQuery`. Since that is currently broken on the client, #6699 should be merged before this one.

diesieben07 commented:
Now that #6699 is merged, this is ready to go.
